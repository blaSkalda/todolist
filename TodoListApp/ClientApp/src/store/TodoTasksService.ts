import { Action, Reducer } from 'redux';
import { AppThunkAction } from '.';

// -----------------
// STATE - This defines the type of data maintained in the Redux store.

export interface TodoTasksState {
    isLoading: boolean;
    isDeleted?: boolean;
    todoTasks: TodoTask[];
    error?: string;
    isCreated?: boolean;
    editedTodoTask?: TodoTask,
}

export interface TodoTask {
    name: string;
    priority: number;
    status: number;
}

// -----------------
// ACTIONS - These are serializable (hence replayable) descriptions of state transitions.
// They do not themselves have any side-effects; they just describe something that is going to happen.

export interface RequestTodoTasksAction {
    type: 'REQUEST_TODO_TASKS';
}

export interface ReceiveTodoTasksAction {
    type: 'RECEIVE_TODO_TASKS';
    todoTasks: TodoTask[];
}

interface CreatedTodoTaskAction {
    type: 'CREATED_TODO_TASK';
    createdTask: TodoTask;
}

interface UpdatedTodoTaskAction {
    type: 'UPDATED_TODO_TASK';
    updatedTask: TodoTask;
    originalName: string;
}

interface EditTodoTaskAction {
    type: 'EDIT_TODO_TASK';
    editTask?: TodoTask;
}

interface DeletedTodoTaskAction {
    type: 'DELETED_TODO_TASK';
    originalName: string;
}

interface ErrorTodoTaskAction {
    type: 'ERROR';
    error: string;
}

interface TodoTaskError {
    message: string;
}

// Declare a 'discriminated union' type. This guarantees that all references to 'type' properties contain one of the
// declared type strings (and not any other arbitrary string).
export type KnownAction = RequestTodoTasksAction | ReceiveTodoTasksAction | CreatedTodoTaskAction
    | ErrorTodoTaskAction | UpdatedTodoTaskAction | DeletedTodoTaskAction
    | EditTodoTaskAction;

// ----------------
// ACTION CREATORS - These are functions exposed to UI components that will trigger a state transition.
// They don't directly mutate state, but they can have external side-effects (such as loading data).

export const actionCreators = {
    requestTodoTasks: (): AppThunkAction<KnownAction> => (dispatch) => {

        dispatch({ type: 'REQUEST_TODO_TASKS' });
        fetch(`api/TodoTasks`)
            .then(response => {
                if (response.status !== 200) {
                    response.json().then(error => {
                        dispatch({ type: 'ERROR', error: error.message });
                    });
                    return;
                }
                return (response.json() as Promise<TodoTask[]>).then(data => {
                    dispatch({ type: 'RECEIVE_TODO_TASKS', todoTasks: data });
                });
            });
    },
    createTodoTask: (task: TodoTask): AppThunkAction<KnownAction> => (dispatch) => {
        dispatch({ type: 'REQUEST_TODO_TASKS' });
        fetch(`api/TodoTasks`, {
            headers: {
                'Content-Type': 'application/json'
            },
            method: 'POST',
            body: JSON.stringify(task)
        })
            .then(response => {
                if (response.status !== 200) {
                    response.json().then(error => dispatch({ type: 'ERROR', error: error.message }));
                    return;
                }
                return (response.json() as Promise<TodoTask>).then(data => {
                    dispatch({ type: 'CREATED_TODO_TASK', createdTask: data });
                });
            });
    },
    updateTodoTask: (name: string, task: TodoTask): AppThunkAction<KnownAction> => (dispatch) => {
        dispatch({ type: 'REQUEST_TODO_TASKS' });
        fetch(`api/TodoTasks/${name}`, {
            headers: {
                'Content-Type': 'application/json'
            },
            method: 'PUT',
            body: JSON.stringify(task)
        })
            .then(response => {
                if (response.status !== 200) {
                    response.json().then(error => dispatch({ type: 'ERROR', error: error.message }));
                    return;
                }
                return (response.json() as Promise<TodoTask>).then(data => {
                    dispatch({ type: 'UPDATED_TODO_TASK', updatedTask: data, originalName: name });
                });
            });
    },
    updateTodoTaskStatus: (name: string, status: number): AppThunkAction<KnownAction> => (dispatch) => {
        dispatch({ type: 'REQUEST_TODO_TASKS' });
        fetch(`api/TodoTasks/${name}/${status}`, {
            headers: {
                'Content-Type': 'application/json'
            },
            method: 'PATCH'
        })
            .then(response => {
                if (response.status !== 200) {
                    response.json().then(error => dispatch({ type: 'ERROR', error: error.message }));
                    return;
                }
                return (response.json() as Promise<TodoTask>).then(data => {
                    dispatch({ type: 'UPDATED_TODO_TASK', updatedTask: data, originalName: name });
                });
            });
    },
    deleteTodoTask: (name: string): AppThunkAction<KnownAction> => (dispatch) => {
        dispatch({ type: 'REQUEST_TODO_TASKS' });
        fetch(`api/TodoTasks/${name}`, {
            headers: {
                'Content-Type': 'application/json'
            },
            method: 'DELETE'
        })
            .then(response => {
                if (response.status !== 200) {
                    response.json().then(error => dispatch({ type: 'ERROR', error: error.message }));
                    return;
                }
                dispatch({ type: 'DELETED_TODO_TASK', originalName: name });
            });
    },
    editTodoTask: (todoTask: TodoTask): AppThunkAction<KnownAction> => (dispatch) => {
        dispatch({ type: 'EDIT_TODO_TASK', editTask: todoTask });
    },
    cancelTodoTask: (): AppThunkAction<KnownAction> => (dispatch) => {
        dispatch({ type: 'EDIT_TODO_TASK', editTask: undefined });
    }
};

// ----------------
// REDUCER - For a given state and action, returns the new state. To support time travel, this must not mutate the old state.

const unloadedState: TodoTasksState = {
    todoTasks: [], isLoading: false
};

export const reducer: Reducer<TodoTasksState> = (state: TodoTasksState | undefined, incomingAction: Action): TodoTasksState => {
    if (state === undefined) {
        return unloadedState;
    }

    const action = incomingAction as KnownAction;
    switch (action.type) {
        case 'REQUEST_TODO_TASKS':
            return {
                todoTasks: state.todoTasks,
                isLoading: true
            };
        case 'RECEIVE_TODO_TASKS':
            return {
                todoTasks: action.todoTasks,
                isLoading: false
            };
        case 'CREATED_TODO_TASK':
            state.todoTasks.push(action.createdTask);
            return {
                todoTasks: Object.assign([], state.todoTasks),
                isCreated: true,
                isLoading: false
            };
        case 'UPDATED_TODO_TASK':
            const updateTaskIndex = state.todoTasks.findIndex((t) => t.name === action.originalName);
            if (updateTaskIndex >= 0)
                state.todoTasks[updateTaskIndex] = action.updatedTask;
            else
                state.todoTasks.push(action.updatedTask);

            return {
                todoTasks: Object.assign([], state.todoTasks),
                isLoading: false,
                editedTodoTask: undefined
            };
        case 'DELETED_TODO_TASK':
            const taskIndex = state.todoTasks.findIndex((t) => t.name === action.originalName);
            if (taskIndex >= 0)
                state.todoTasks.splice(taskIndex, 1);

            return {
                todoTasks: Object.assign([], state.todoTasks),
                isDeleted: true,
                isLoading: false
            };
        case 'EDIT_TODO_TASK':
            return {
                todoTasks: state.todoTasks,
                editedTodoTask: action.editTask,
                isLoading: false
            };
        case 'ERROR':
            return {
                error: action.error,
                todoTasks: state.todoTasks,
                editedTodoTask: undefined,
                isLoading: false
            };
    }

    return state;
};
