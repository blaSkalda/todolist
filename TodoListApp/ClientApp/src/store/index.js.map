{"version":3,"file":"index.js","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":";;;AAAA,qDAAuD;AAOvD,sGAAsG;AACtG,wGAAwG;AACxG,4DAA4D;AAC/C,QAAA,QAAQ,GAAG;IACpB,SAAS,EAAE,gBAAgB,CAAC,OAAO;CACtC,CAAC"}