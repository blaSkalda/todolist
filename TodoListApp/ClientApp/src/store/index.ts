import * as TodoTasksService from './TodoTasksService';

// The top-level state object
export interface ApplicationState {
    todoTasks: TodoTasksService.TodoTasksState | undefined;
}

// Whenever an action is dispatched, Redux will update each top-level application state property using
// the reducer with the matching name. It's important that the names match exactly, and that the reducer
// acts on the corresponding ApplicationState property type.
export const reducers = {
    todoTasks: TodoTasksService.reducer
};
