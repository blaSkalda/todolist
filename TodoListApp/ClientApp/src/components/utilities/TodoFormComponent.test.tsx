import * as React from 'react';
import { shallow } from 'enzyme'

import TodoForm, { TodoFormState } from './TodoFormComponent';
import { TodoTask } from '../../store/TodoTasksService';

describe('<TodoForm />', () => {
    const props = {
        editedTodoTask: undefined as TodoTask | undefined,
        isLoading: false,
        todoTasks: [],
        todoTask: {
            name: 'Test',
            priority: 1,
            status: 1
        },
        updateTodoTask: jest.fn(),
        cancelTodoTask: jest.fn(),
        editTodoTask: jest.fn(),
        deleteTodoTask: jest.fn(),
        updateTodoTaskStatus: jest.fn(),
        createTodoTask: jest.fn(),
        requestTodoTasks: jest.fn(),
    };

    it('renders <TodoForm />', () => {
        const wrapper = shallow(<TodoForm {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('renders <TodoForm /> and changes state to creating new todo list', () => {
        const wrapper = shallow(<TodoForm {...props} />);
        wrapper.setState({ isCreating: true });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders <TodoForm /> in edit state', () => {
        props.editedTodoTask = {
            name: 'Test'
        } as TodoTask;
        const wrapper = shallow(<TodoForm {...props} />);

        expect(wrapper).toMatchSnapshot();
    });


    it('renders <TodoForm /> and changes state to creating new todo list adn changes name for new todo task', () => {
        props.editedTodoTask = undefined;

        const wrapper = shallow(<TodoForm {...props} />);
        wrapper.setState({ isCreating: true });

        wrapper.find('#name').simulate('change', { target: { value: 'Skalda', id: 'name' } });

        const state = wrapper.state() as TodoFormState;

        expect(state.todoTask.name).toEqual('Skalda');
    });

    it('renders <TodoForm /> edit todo list adn changes name for new todo task', () => {
        props.editedTodoTask = {
            name: 'Test'
        } as TodoTask;

        const wrapper = shallow(<TodoForm {...props} />);
        wrapper.setState({ isCreating: true });

        wrapper.find('#name').simulate('change', { target: { value: 'Skalda', id: 'name'} });

        const state = wrapper.state() as TodoFormState;

        expect(state.todoTask.name).toEqual('Skalda');
    });
});
