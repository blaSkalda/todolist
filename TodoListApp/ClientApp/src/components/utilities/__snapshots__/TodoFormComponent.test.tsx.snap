// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoForm /> renders <TodoForm /> 1`] = `
<button
  className="btn btn-primary"
  onClick={[Function]}
>
  Create
</button>
`;

exports[`<TodoForm /> renders <TodoForm /> and changes state to creating new todo list 1`] = `
<div
  className="jumbotron"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="name"
      >
        Task name
      </label>
      <input
        className="form-control"
        id="name"
        onChange={[Function]}
        placeholder="Todo task name"
        type="text"
        value="Test"
      />
      <small
        className="form-text text-muted"
      >
        Your TODO task name
      </small>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="priority"
      >
        Priority
      </label>
      <input
        className="form-control"
        id="priority"
        onChange={[Function]}
        placeholder="Priority"
        type="number"
        value={1}
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="status"
      >
        Status
      </label>
      <select
        className="form-control"
        id="status"
        onChange={[Function]}
        value={1}
      >
        <option
          key="0"
          value="0"
        >
          Not started
        </option>
        <option
          key="1"
          value="1"
        >
          In progress
        </option>
        <option
          key="2"
          value="2"
        >
          Completed
        </option>
      </select>
    </div>
    <div
      className="d-flex justify-content-between"
    >
      <button
        className="btn btn-primary"
        type="submit"
      >
        Create new task
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        type="submit"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`<TodoForm /> renders <TodoForm /> in edit state 1`] = `
<div
  className="jumbotron"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="name"
      >
        Task name
      </label>
      <input
        className="form-control"
        id="name"
        onChange={[Function]}
        placeholder="Todo task name"
        type="text"
        value="Test"
      />
      <small
        className="form-text text-muted"
      >
        Your TODO task name
      </small>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="priority"
      >
        Priority
      </label>
      <input
        className="form-control"
        id="priority"
        onChange={[Function]}
        placeholder="Priority"
        type="number"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="status"
      >
        Status
      </label>
      <select
        className="form-control"
        id="status"
        onChange={[Function]}
      >
        <option
          key="0"
          value="0"
        >
          Not started
        </option>
        <option
          key="1"
          value="1"
        >
          In progress
        </option>
        <option
          key="2"
          value="2"
        >
          Completed
        </option>
      </select>
    </div>
    <div
      className="d-flex justify-content-between"
    >
      <button
        className="btn btn-primary"
        type="submit"
      >
        Update task
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        type="submit"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
